File: dotnet/src/Microsoft.Agents.Validation/Validators/MessageValidators/SystemMessageValidator.cs
using Microsoft.Agents.Abstractions.Models;

namespace Microsoft.Agents.Validation.Validators.MessageValidators;

/// <summary>
/// Validator for SystemMessage.
/// </summary>
public class SystemMessageValidator : MessageValidatorBase<SystemMessage>
{
    protected override List<ValidationError> ValidateSpecificFields(SystemMessage message, ValidationContext? context)
    {
        var errors = new List<ValidationError>();

        // MSG-008: SystemMessage content must be non-empty
        if (string.IsNullOrWhiteSpace(message.Content))
        {
            errors.Add(CreateError(
                ValidationErrorCode.MSG_008,
                "SystemMessage content must be non-empty",
                "Content"));
        }

        // ROLE-001: SystemMessage can only contain TextContent
        // Note: SystemMessage uses Content property, not Contents list
        // This rule applies when SystemMessage is extended to support Contents

        return errors;
    }
}

File: typescript/packages/agents-protocol-validation/src/validators/message/SystemMessageValidator.ts
import type { SystemMessage } from '@microsoft/agents-protocol-abstractions';
import type { ValidationContext } from '../../core/ValidationContext';
import { MessageValidatorBase } from '../base/MessageValidatorBase';
import { ValidationError } from '../../core/ValidationError';
import { ValidationErrorCode } from '../../core/ValidationErrorCode';

/**
 * Validator for SystemMessage.
 */
export class SystemMessageValidator extends MessageValidatorBase<SystemMessage> {
  protected validateSpecificFields(message: SystemMessage, context?: ValidationContext): ValidationError[] {
    // MSG-008: System messages have no special fields beyond common ones
    return [];
  }

  protected getAllowedContentTypes(): string[] {
    // ROLE-001: System messages can only contain text, document, adaptive-card
    return ['text', 'document', 'adaptive-card'];
  }

  protected getRoleSpecificErrorCode(): string {
    return ValidationErrorCode.ROLE_001;
  }

  protected getMessageTypeName(): string {
    return 'SystemMessage';
  }

  protected requiresContent(): boolean {
    return false; // System messages can be empty
  }
}
